pkg/binlog/event/util.go

```diff
@@ -0,0 +1,270 @@
+			pos += 2
+		case gmysql.MYSQL_TYPE_BLOB, gmysql.MYSQL_TYPE_DOUBLE, gmysql.MYSQL_TYPE_FLOAT, gmysql.MYSQL_TYPE_GEOMETRY, gmysql.MYSQL_TYPE_JSON,
+			gmysql.MYSQL_TYPE_TIME2, gmysql.MYSQL_TYPE_DATETIME2, gmysql.MYSQL_TYPE_TIMESTAMP2:
+			columnMeta[i] = uint16(data[pos])
+			pos++
+		case gmysql.MYSQL_TYPE_NEWDATE, gmysql.MYSQL_TYPE_ENUM, gmysql.MYSQL_TYPE_SET, gmysql.MYSQL_TYPE_TINY_BLOB, gmysql.MYSQL_TYPE_MEDIUM_BLOB, gmysql.MYSQL_TYPE_LONG_BLOB:
+			return nil, errors.NotSupportedf("column type %d in binlog", t)
+		default:
+			columnMeta[i] = 0
+		}
+	}
+
+	return columnMeta, nil
+}
+
+// bitmapByteSize returns the byte length of bitmap for columnCount.
+func bitmapByteSize(columnCount int) int {
+	return int(columnCount+7) / 8
+}
+
+// nullBytes returns a n-length null bytes slice
+func nullBytes(n int) []byte {
+	buf := new(bytes.Buffer)
+	for i := 0; i < n; i++ {
+		buf.WriteByte(0x00)
+	}
+	return buf.Bytes()
+}
+
+// fullBytes returns a n-length full bytes slice (all bits are set)
+func fullBytes(n int) []byte {
+	buf := new(bytes.Buffer)
+	for i := 0; i < n; i++ {
+		buf.WriteByte(0xff)
+	}
+	return buf.Bytes()
+}
+
+// assembleEvent assembles header fields, postHeader and payload together to an event.
+// header: pass as a struct to make a copy
+func assembleEvent(buf *bytes.Buffer, event replication.Event, decodeWithChecksum bool, header replication.EventHeader, eventType replication.EventType, latestPos uint32, postHeader, payload []byte) (*replication.BinlogEvent, error) {
+	eventSize := uint32(eventHeaderLen) + uint32(len(postHeader)) + uint32(len(payload)) + crc32Len
+	// update some fields in header
+	header.EventSize = eventSize
+	header.LogPos = latestPos + eventSize
+	header.EventType = eventType
+	headerData, err := GenEventHeader(&header)
+	if err != nil {
+		return nil, errors.Annotatef(err, "generate event header")
+	}
+
+	err = combineHeaderPayload(buf, headerData, postHeader, payload)
+	if err != nil {
+		return nil, errors.Annotatef(err, "combine header, post-header and payload")
+	}
+
+	// CRC32 checksum, 4 bytes
+	checksum := crc32.ChecksumIEEE(buf.Bytes())
+	err = binary.Write(buf, binary.LittleEndian, checksum)
+	if err != nil {
+		return nil, errors.Annotatef(err, "write CRC32 % X", checksum)
+	}
+
+	if event == nil {
+		return nil, nil // not need to decode the event
+	}
+
+	// decode event, some implementations of `Decode` also need checksum
+	var endIdx = buf.Len()
+	if !decodeWithChecksum {
+		endIdx -= int(crc32Len)
+	}
+	err = event.Decode(buf.Bytes()[eventHeaderLen:endIdx])
+	if err != nil {
+		return nil, errors.Annotatef(err, "decode % X", buf.Bytes())
+	}
+
+	return &replication.BinlogEvent{RawData: buf.Bytes(), Header: &header, Event: event}, nil
+}
+
+// combineHeaderPayload combines header, postHeader and payload together.
+func combineHeaderPayload(buf *bytes.Buffer, header, postHeader, payload []byte) error {
+	err := binary.Write(buf, binary.LittleEndian, header)
```

Comment:
do you want check `its length == eventHeaderLen`?

Reply:
added